File: Data Structures & Algorithms/edit-distance/submission-3.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n, m = len(word1), len(word2)
        dp = {}

        def dfs(i, j):
            if j == m:
                return n-i
            if i == n:
                return m-j
            if (i, j) in dp:
                return dp[(i, j)]
            if word1[i] == word2[j]:
                dp[(i, j)] = dfs(i+1, j+1)
            else:
                res = min(dfs(i, j+1), dfs(i+1, j))
                res = min(res, dfs(i+1, j+1))
                dp[(i, j)] = res+1
            return dp[(i, j)]
        
        return dfs(0,0)

File: Data Structures & Algorithms/jump-game/submission-3.java
class Solution {
    public boolean canJump(int[] nums) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp, -1);
        return dfs(nums, 0, dp);
    }
    boolean dfs(int[] nums, int i, int[] dp){
        if(i == nums.length-1){
            return true;
        }
        if(nums[i] == 0){
            return false;
        }
        if(dp[i] != -1){
            return dp[i] == 1 ? true : false;
        }
        int end = Math.min(nums.length-1, i + nums[i]);
        for(int j=i+1; j<=end; j++){
            if(dfs(nums, j, dp)){
                dp[i] = 1;
                return true;
            }
        }
        dp[i] = 0;
        return false;
    }
}
